Update deployment documentation (#10882)
dask/dask: docs/source/deploying.rst

Cloud
-----

> Deploying on commercial cloud like AWS, GCP, or Azure is convenient because you can quickly scale out to many machines for just a few minutes, but also challenging because you need to navigate awkward cloud APIs, manage remote software environments with Docker, send data access credentials, make sure that costly resources are cleaned up, etc..  The following solutions help with this process.

Reviewer: I don't think we need the extra `.` in this case

```suggestion
Deploying on commercial cloud like AWS, GCP, or Azure is convenient because you can quickly scale out to many machines for just a few minutes, but also challenging because you need to navigate awkward cloud APIs, manage remote software environments with Docker, send data access credentials, make sure that costly resources are cleaned up, etc.  The following solutions help with this process.
```

-   `Coiled (recommended) <https://coiled.io?utm_source=dask-docs&utm_medium=deploying>`_:

>     this commercial SaaS product handles most of the deployment pain we currently
>     see, is easy to use, and quite robust.  The free tier is generous enough

Reviewer: [Nit] Slight movement away from "we" towards end users instead

```suggestion
    this commercial SaaS product handles most of the deployment pain Dask users
    encounter, is easy to use, and quite robust.  The free tier is generous enough
```

>     this commercial SaaS product handles most of the deployment pain we currently
>     see, is easy to use, and quite robust.  The free tier is generous enough

Reviewer: ```suggestion
    this commercial SaaS product handles most of the deployment pain we currently
    see, is easy to use, and quite robust.  The free tier is large enough
```

    for most individual users, even for those who don't want to engage with a
    commercial company.  The API looks like the following.

    .. code-block:: python

       import coiled
       cluster = coiled.Cluster(
           n_workers=100,
           region="us-east-2",
           worker_memory="16 GiB",
           spot_policy="spot_with_fallback",
       )
       client = cluster.get_client()

- `Dask Cloud Provider <https://cloudprovider.dask.org/en/latest/>`_: a pure and simple OSS solution that sets up Dask workers on cloud VMs, supporting AWS, GCP, Azure, and also other commercial clouds like Hetzner and Digital Ocean.

> - `Dask-Yarn <https://yarn.dask.org>`_: deploys Dask on legacy YARN clusters, such as can be set up with AWS EMR or Google Cloud Dataproc

Reviewer: [Non-blocking comment] You have a much better sense for if the term "legacy" applies here. I definitely see less Dask traffic around YARN clusters these days -- just noting that this is the first time I'm seeing the term "legacy" in this context.

Author: Yeah, I'm comfortable throwing some shade at Yarn today.

> - `Dask-Yarn <https://yarn.dask.org>`_: deploys Dask on legacy YARN clusters, such as can be set up with AWS EMR or Google Cloud Dataproc

Reviewer: ```suggestion
- `Dask-Yarn <https://yarn.dask.org>`_: deploys Dask on legacy YARN clusters, such as can be set up with AWS EMR or Google Cloud Dataproc.
```

See :doc:`deploying-cloud` for more details.

.. _Coiled: https://coiled.io?utm_source=dask-docs&utm_medium=deploying

> .. |Coiled| replace:: **Coiled**

Reviewer: This (and the similar lines in other sections) aren't needed anymore if we're not going to bold the recommended deployment project. 

Thoughts on continuing to bold them?

Author: Happy to bold them, I just didn't want to deal with the sphinx stuff to make that happen.

Reviewer: No problem -- just pushed a commit that makes them bold

High Performance Computing
--------------------------

Dask runs on traditional HPC systems that use a resource manager like SLURM,
PBS, SGE, LSF, or similar systems, and a network file system.  This is an easy
way to dual-purpose large-scale hardware for analytics use cases.  Dask can
deploy either directly through the resource manager or through
``mpirun``/``mpiexec`` and tends to use the NFS to distribute data and
software.

-   `Dask-Jobqueue (recommended) <https://jobqueue.dask.org>`_: interfaces directly with the
    resource manager (SLURM, PBS, SGE, LSF, and others) to launch many Dask
    workers as batch jobs.  It generates batch job scripts and submits them
    automatically to the user's queue.  This approach operates entirely with user
    permissions (no IT support required) and enables interactive and adaptive use
    on large HPC systems.  It looks a little like the following:

    .. code-block:: python

>        from dask_jobqueue import PBSCluster
> 
>        cluster = PBSCluster(
>            cores=24,
>            memory="100GB",
>            queue="regular",
>            account="my-account",
>        )
>        cluster.scale(jobs=100)
> 
>        client = cluster.get_client()

Reviewer: [Nit] Removing extra spaces to match the other example snippets (or add extra spaces to the other example) 

```suggestion
       from dask_jobqueue import PBSCluster
       cluster = PBSCluster(
           cores=24,
           memory="100GB",
           queue="regular",
           account="my-account",
       )
       cluster.scale(jobs=100)
       client = cluster.get_client()
```

- `Dask-MPI <http://mpi.dask.org/en/latest/>`_: deploys Dask on top of any system that supports MPI using ``mpirun``.  It is helpful for batch processing jobs where you want to ensure a fixed and stable number of workers.
- `Dask Gateway for Jobqueue <https://gateway.dask.org/install-jobqueue.html>`_: Multi-tenant, secure clusters. Once configured, users can launch clusters without direct access to the underlying HPC backend.

See :doc:`deploying-hpc` for more details.

.. _Dask-Jobqueue: https://jobqueue.dask.org
.. |Dask-Jobqueue| replace:: **Dask-Jobqueue**

Kubernetes
----------

Dask runs natively on Kubernetes clusters.  This is a convenient choice when a
company already has dedicated Kubernetes infrastructure set up for running
other services.  When running Dask on Kubernetes users should also have a plan

> to distribute software environments (probably with Docker) user credentials,

Reviewer: ```suggestion
to distribute software environments (probably with Docker), user credentials,
```

Reviewer: Is there an option outside of Docker?

> quota management, etc.. In larger organizations with mature Kubernetes

Reviewer: ```suggestion
quota management, etc. In larger organizations with mature Kubernetes
```

deployments this is often handled by other Kubernetes services.

-   `Dask Kubernetes Operator (recommended)
    <https://kubernetes.dask.org/en/latest/operator.html>`_: The Dask Kubernetes
    Operator makes the most sense for fast moving or ephemeral deployments.  It
    is the most Kubernetes-native solution, and should be comfortable for K8s
    enthusiasts.  It looks a little like this:

    .. code-block:: python

>        from dask_kubernetes.operator import KubeCluster
>        cluster = KubeCluster(
>            name="my-dask-cluster",
>            image='ghcr.io/dask/dask:latest',
>            resources={"requests": {"memory": "2Gi"}, "limits": {"memory": "64Gi"}},
>        )
>        cluster.scale(10)
> 
>        client = cluster.get_client()

Reviewer: ```suggestion
       from dask_kubernetes.operator import KubeCluster
       cluster = KubeCluster(
           name="my-dask-cluster",
           image='ghcr.io/dask/dask:latest',
           resources={"requests": {"memory": "2Gi"}, "limits": {"memory": "64Gi"}},
       )
       cluster.scale(10)
       client = cluster.get_client()
```
